packages/fuels-contract/src/predicate.rs

```diff
@@ -0,0 +1,136 @@
+use crate::script::Script;
+use anyhow::Result;
+use fuel_gql_client::{
+    fuel_tx::{Contract, Input, Output, Receipt, Transaction, UtxoId},
+    fuel_types::{Address, AssetId},
+};
+
+use fuel_gql_client::fuel_vm::prelude::Opcode;
+use fuels_core::parameters::TxParameters;
+use fuels_signers::{provider::Provider, wallet::Wallet, Signer};
+use fuels_types::{bech32::Bech32Address, errors::Error};
+
+/// Predicate provides methods to create new predicates and call them
+pub struct Predicate {
+    pub address: Address,
+    code: Vec<u8>,
+}
+
+impl Predicate {
+    pub fn new(code: Vec<u8>) -> Self {
+        let address = (*Contract::root_from_code(&code)).into();
+        Self { code, address }
+    }
+
+    /// Craft a transaction used to transfer funds between two addresses.
+    pub fn build_transfer_tx(
+        &self,
+        inputs: &[Input],
+        outputs: &[Output],
+        params: TxParameters,
+    ) -> Transaction {
+        // This script contains a single Opcode that returns immediately (RET)
+        // since all this transaction does is move Inputs and Outputs around.
+        let script = Opcode::RET(0x10).to_bytes().to_vec();
+        Transaction::Script {
+            gas_price: params.gas_price,
+            gas_limit: params.gas_limit,
+            byte_price: params.byte_price,
+            maturity: params.maturity,
+            receipts_root: Default::default(),
+            script,
+            script_data: vec![],
+            inputs: inputs.to_vec(),
+            outputs: outputs.to_vec(),
+            witnesses: vec![],
+            metadata: None,
+        }
+    }
+
+    /// Deploys locked coins in a Predicate to the given wallet's provider.
+    ///
+    /// # Arguments
+    ///
+    /// * `wallet` - A wallet that funds this transaction
+    /// * `coin_amount_to_predicate` - The amount of locked coins as given asset id to store within Predicate
+    /// * `asset_id` - The asset id of the locked coins stored within Predicate
+    pub async fn deploy_predicate(
```

Comment:
I dislike that we are duplicating code here from `wallet.transfer()`.

Maybe we could either move this to `wallet` and have it  be a wrapper for `wallet.transfer()` or extract `wallet.transfer()`s functionality into a provider method and have `wallet` and `predicate` use it. 
`spend_predicate` from below could also benefit from this.

This can also be done in a separate refactoring issue, just wanted to share my thoughts.

Reply:
I would suggest a separate issue that someone from the SDK team takes - I would volunteer.  We also need to add documentation as I have not seen an open issuer for that.